envoy/rds/config_traits.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+#include "envoy/rds/config.h"
+
+#include "source/common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Rds {
+
+/**
+ * Traits of the protocol specific route configuration and proto.
+ * The generic rds classes will call the methods of this interface
+ * to get information which are not visible for them directly.
+ */
+class ConfigTraits {
+public:
+  virtual ~ConfigTraits() = default;
+
+  /**
+   * Give the full name of the route configuration proto description.
+   * For example 'envoy.config.route.v3.RouteConfiguration'
+   */
+  virtual std::string resourceType() const PURE;
+
+  /**
+   * Create a dummy config object without actual route configuration.
+   * This object will be used before the first valid route configuration is fetched.
+   */
+  virtual ConfigConstSharedPtr createConfig() const PURE;
+
+  /**
+   * Create an empty route configuration proto object.
+   */
+  virtual ProtobufTypes::MessagePtr createProto() const PURE;
+
+  /**
+   * Runtime check if the provided proto message object is really a route configuration instance.
+   * Throw an std::bad_cast exception if not.
+   * Every other method below this assumes the proto message is already
+   * validated and doesn't do any further runtime check.
+   */
+  virtual const Protobuf::Message& validateResourceType(const Protobuf::Message& rc) const PURE;
+
+  /**
+   * Gives back the name from the route configuration proto.
+   * The object behind the returned reference has to have the same lifetime like the proto.
+   */
+  virtual const std::string& resourceName(const Protobuf::Message& rc) const PURE;
+
+  /**
+   * Create a config object based on a route configuration.
+   * @throw EnvoyException if the new config can't be applied.
+   */
+  virtual ConfigConstSharedPtr createConfig(const Protobuf::Message& rc) const PURE;
+
+  /**
+   * Clones the route configuration proto.
+   */
+  virtual ProtobufTypes::MessagePtr cloneProto(const Protobuf::Message& rc) const PURE;
```

Comment:
I'm thinking this is a wider surface than needed. Happy to be persuaded otherwise, but I think we should make it easy to get at the proto (by reference if needed) and then just use generic clone on that (or via reflection if needed). One thing we could thing about is using dynamic reflection for proto here to avoid needing to navigate typing shenanigans. Reflection can simplify without overhead if we're not doing crazy traversal stuff like we did in the upgrader.

Reply:
What do you think about just delete this function form the interface an where it is used replace to:
p = createEmptyProto()
p->CopyFrom(rc);
I think it can work.

Reply:
During the change I realized the interface is really too wide. So I plan to split:
`ConfigTraits:
createNullConfig()
createConfig(rc)

ProtoTraits:
resourceType()
createEmptyProto()
validateResourceType(rc)
resourceName(rc)`

The `ProtoTraits` doesn't require extra parameters for sure like `OptionalHttpFilters` at `ConfigTraits`. Enough to instantiate once in `Router::RouteConfigProviderManagerImpl` and pass as reference wherever it is needed. At least I can revert the newly added `ConfigInfo::name_` field.

We could replace the call for `validateResourceType` also by simply compare the rc.GetDescriptor()->full_name() and proto_traits_.resourceType(). My only concern with this is in general not true if the full name is matching then the dynamic cast works. But maybe in case of envoy they are equivalent.

If we stay at the type validation by dynamic cast I would also put back the `cloneProto` function. The copy constructor maybe more effective than the `CopyFrom`.

Which validation method do you prefer, message full name comparison or the dynamic cast?

Reply:
I like the new structure very much, this is pretty orthogonal and minimal. I probably wouldn't argue too much against `cloneProto`, but TBH, if you are cloning a route table (potentially large data structure), you've already lost the efficiency battle, so I'm not sure it's worth adding directly the interface.

I hear what you are saying about protos. In general, you can't assume it's safe to cast based just on type name equivalence, since you might have constructed a proto entirely using reflection for example, e.g. with https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.dynamic_message#DynamicMessageFactory.

However, I think for the purposes of RDS reuse here, it's all going to be concrete instances of the protos and this will be safe. We can add the complexity later if things change.

Reply:
Ok, I have removed the `cloneProto`.

I have also simplified the `ProtoTraits` further by removing the `validateResourceType`.
I have the same understanding like you: in the scope of rds the type name match means the dynamic cast will succeed.
Anyway if because some reason the type name matches but the underlying class is different it won't cause problem:
- In `Rds::RouteConfigUpdateReceiverImpl::onRdsUpdate` a cloned proto is stored, That has the right class for sure and the casts later happen on this.
- In `Router::RouteConfigUpdateReceiverImpl::onRdsUpdate` I changed a the previous static cast to dynamic, so the overall working will be the same like with the `validateResourceType` function.